hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java

```diff
@@ -150,29 +160,109 @@ public static void parseKVContainerData(KeyValueContainerData kvContainerData,
+            new MetadataKeyFilters.KeyPrefixFilter()
+                .addFilter(OzoneConsts.DELETING_KEY_PREFIX);
+        int numPendingDeletionBlocks =
+            containerDB.getStore().getSequentialRangeKVs(null,
+                Integer.MAX_VALUE, filter)
+                .size();
+        kvContainerData.incrPendingDeletionBlocks(numPendingDeletionBlocks);
+      }
+
+      // Set delete transaction id.
+      byte[] delTxnId =
+          containerDB.getStore().get(DB_CONTAINER_DELETE_TRANSACTION_KEY);
+      if (delTxnId != null) {
+        kvContainerData
+            .updateDeleteTransactionId(Longs.fromByteArray(delTxnId));
+      }
+
+      // Set BlockCommitSequenceId.
+      byte[] bcsId = containerDB.getStore().get(
+          DB_BLOCK_COMMIT_SEQUENCE_ID_KEY);
+      if (bcsId != null) {
+        kvContainerData
+            .updateBlockCommitSequenceId(Longs.fromByteArray(bcsId));
+      }
+
+      // Set bytes used.
+      // commitSpace for Open Containers relies on usedBytes
+      byte[] bytesUsed =
+          containerDB.getStore().get(DB_CONTAINER_BYTES_USED_KEY);
+      if (bytesUsed != null) {
+        isBlockMetadataSet = true;
+        kvContainerData.setBytesUsed(Longs.fromByteArray(bytesUsed));
+      }
+
+      // Set block count.
+      byte[] blockCount = containerDB.getStore().get(DB_BLOCK_COUNT_KEY);
+      if (blockCount != null) {
+        isBlockMetadataSet = true;
+        kvContainerData.setKeyCount(Longs.fromByteArray(blockCount));
+      }
+    }
+
+    if (!isBlockMetadataSet) {
+      initializeUsedBytesAndBlockCount(kvContainerData);
+    }
+  }
+
+
+  /**
+   * Initialize bytes used and block count.
+   * @param kvContainerData
+   * @throws IOException
+   */
+  private static void initializeUsedBytesAndBlockCount(
+      KeyValueContainerData kvContainerData) throws IOException {
+
+    long blockCount = 0;
+    try (KeyValueBlockIterator blockIter = new KeyValueBlockIterator(
+        kvContainerData.getContainerID(),
+        new File(kvContainerData.getContainerPath()))) {
+      long usedBytes = 0;
+
+
+      boolean success = true;
+      while (success) {
         try {
-          blockData = BlockUtils.getBlockData(e.getValue());
-          return blockData.getSize();
+          if (blockIter.hasNext()) {
+            BlockData block = blockIter.nextBlock();
+            long blockLen = 0;
+
+            List< ContainerProtos.ChunkInfo > chunkInfoList = block.getChunks();
+            for (ContainerProtos.ChunkInfo chunk : chunkInfoList) {
+              ChunkInfo info = ChunkInfo.getFromProtoBuf(chunk);
+              blockLen += info.getLen();
+            }
+
+            usedBytes += blockLen;
+            blockCount++;
+          } else {
+            success = false;
+          }
         } catch (IOException ex) {
-          return 0L;
+          LOG.error("Failed to parse block data for Container {}",
+              kvContainerData.getContainerID());
```

Comment:
Previously also we used to do this. Slightly modified the code. And I think for one block corruption, if we don't add it to container set, then BackGround Container Scanner will never know about this container, and it will never be fixed. So, I think adding to container set is right thing to do here. Let me know your thoughts?

```
     bytesUsed = liveKeys.parallelStream().mapToLong(e-> {
        BlockData blockData;
        try {
          blockData = BlockUtils.getBlockData(e.getValue());
          return blockData.getSize();
        } catch (IOException ex) {
          return 0L;
        }
      }).sum();
```

Reply:
I see that previously initializeUsedBytes() was done only for OPEN containers. And for others, we would just get the BlockData and update usedBytes.  
What I mean to say is that the behavior has changed with this patch. I might have missed some background here. Can you please explain why the change.

Reply:
Previously we used to compute twice, one using different logic and 2nd using different logic. Now we are doing only once by using initializeUsedBytesAndBlockCount logic.

Reply:
And also at the end, even new logic which is using iterator finally uses BlockUtils.getBlockData to getNext Block. I feel the current way of iterating and computing the size is cleaner way, instead of getting all blocks into in-memory at once.

Reply:
Thanks for the explanation Bharat. Makes sense.